op-service/dial/active_rollup_provider.go

```diff
@@ -0,0 +1,169 @@
+	networkTimeout time.Duration,
+	logger log.Logger,
+) (*ActiveL2RollupProvider, error) {
+	rollupDialer := func(ctx context.Context, timeout time.Duration,
+		log log.Logger, url string) (RollupClientInterface, error) {
+		return DialRollupClientWithTimeout(ctx, timeout, log, url)
+	}
+	return newActiveL2RollupProvider(ctx, rollupUrls, checkDuration, networkTimeout, logger, rollupDialer)
+}
+
+func newActiveL2RollupProvider(
+	ctx context.Context,
+	rollupUrls []string,
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+	dialer rollupDialer,
+) (*ActiveL2RollupProvider, error) {
+	if len(rollupUrls) == 0 {
+		return nil, errors.New("empty rollup urls list")
+	}
+	p := &ActiveL2RollupProvider{
+		checkDuration:  checkDuration,
+		networkTimeout: networkTimeout,
+		log:            logger,
+		rollupUrls:     rollupUrls,
+		rollupDialer:   dialer,
+		clientLock:     &sync.Mutex{},
+		rollupIndex:    -1,
+	}
+	cctx, cancel := context.WithTimeout(ctx, networkTimeout)
+	defer cancel()
+	err := p.ensureClientInitialized(cctx)
+	if err != nil {
+		return nil, fmt.Errorf("dialing initial rollup client: %w", err)
+	}
+	_, err = p.RollupClient(cctx)
+	if err != nil {
+		return nil, fmt.Errorf("setting provider rollup client: %w", err)
+	}
+	return p, nil
+}
+
+var errSeqUnset = errors.New("sequencer unset")
+
+func (p *ActiveL2RollupProvider) RollupClient(ctx context.Context) (RollupClientInterface, error) {
+	p.clientLock.Lock()
+	defer p.clientLock.Unlock()
+	err := p.ensureActiveEndpoint(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return p.currentRollupClient, nil
+}
+
+func (p *ActiveL2RollupProvider) ensureActiveEndpoint(ctx context.Context) error {
+	if !p.shouldCheck() {
+		return nil
+	}
+	if err := p.findActiveEndpoints(ctx); err != nil {
+		return err
+	}
+	p.activeTimeout = time.Now().Add(p.checkDuration)
+	return nil
+}
+
+func (p *ActiveL2RollupProvider) shouldCheck() bool {
+	return time.Now().After(p.activeTimeout)
+}
+
+func (p *ActiveL2RollupProvider) ensureClientInitialized(ctx context.Context) error {
+	if p.currentRollupClient != nil {
+		return nil
+	}
+	return p.dialNextSequencer(ctx)
+}
+
+func (p *ActiveL2RollupProvider) findActiveEndpoints(ctx context.Context) error {
+	for range p.rollupUrls {
+		active, err := p.checkCurrentSequencer(ctx)
+		if errors.Is(err, errSeqUnset) {
+			log.Debug("Current sequencer unset.")
+		} else if ep := p.rollupUrls[p.rollupIndex]; err != nil {
+			p.log.Warn("Error querying active sequencer, closing connection and trying next.", "err", err, "index", p.rollupIndex, "url", ep)
+		} else if active {
+			p.log.Debug("Current sequencer active.", "index", p.rollupIndex, "url", ep)
+			return nil
+		} else {
+			p.log.Info("Current sequencer inactive, closing connection and trying next.", "index", p.rollupIndex, "url", ep)
+		}
+		if err := p.dialNextSequencer(ctx); err != nil {
+			return fmt.Errorf("dialing next sequencer: %w", err)
+		}
```

Comment:
I think we want to do
```suggestion
		if err := p.dialNextSequencer(ctx); err != nil {
			p.log.Warn("Error dialing next sequencer.", "err", err, "index", p.rollupIndex)
		}
```
here and exhaustively iterating over the whole loop. If dialing fails, we also want to try the next sequencer.

I think if we do that, we might be able to remove `ensureClientInitialized` and also not dial from the ctor.